cassandane/Cassandane/Cyrus/Alpaca.pm

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/perl
+#
+#  Copyright (c) 2011-2024 FastMail Pty Ltd. All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions
+#  are met:
+#
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in
+#     the documentation and/or other materials provided with the
+#     distribution.
+#
+#  3. The name "Fastmail Pty Ltd" must not be used to
+#     endorse or promote products derived from this software without
+#     prior written permission. For permission or any legal
+#     details, please contact
+#      FastMail Pty Ltd
+#      PO Box 234
+#      Collins St West 8007
+#      Victoria
+#      Australia
+#
+#  4. Redistributions of any form whatsoever must retain the following
+#     acknowledgment:
+#     "This product includes software developed by Fastmail Pty. Ltd."
+#
+#  FASTMAIL PTY LTD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
+#  INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY  AND FITNESS, IN NO
+#  EVENT SHALL OPERA SOFTWARE AUSTRALIA BE LIABLE FOR ANY SPECIAL, INDIRECT
+#  OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
+#  USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+#  TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
+#  OF THIS SOFTWARE.
+#
+
+package Cassandane::Cyrus::Alpaca;
+use strict;
+use warnings;
+use Data::Dumper;
+
+use lib '.';
+use base qw(Cassandane::Cyrus::TestCase);
+use Cassandane::Util::Log;
+use Cassandane::Util::Socket;
+
+sub new
+{
+    my $class = shift;
+    return $class->SUPER::new({}, @_);
+}
+
+sub set_up
+{
+    my ($self) = @_;
+    $self->SUPER::set_up();
+}
+
+sub tear_down
+{
+    my ($self) = @_;
+    $self->SUPER::tear_down();
+}
+
+sub imap_cmd_with_tag
+{
+    my ($talk, $tag, @args) = @_;
+
+    die "not a Mail::IMAPTalk object" if ref $talk ne 'Mail::IMAPTalk';
+
+    # save some internal state
+    my $saved_CmdId = $talk->{CmdId};
+
+    # override next tag with the tag we want to use
+    $talk->{CmdId} = $tag;
```

Comment:
Interesting.  I haven't done that sort of thing much, mostly only for perl internal stuff like $SIG, $/, and so on.  Would I do that like `local $talk->{CmdId} = $tag;`?

Smells like there could be dragons here because this is a hashref rather than a hash.  There probably isn't, perl will probably do what I mean, but it has a smell about it.

Reply:
Yes that's how you'd do it.

It's the same mechanism, `%SIG` is just a special hash. `local` is special for hash values when used this way.

Anything in the current block and any deeper blocks will see the localized value, and the real value will be restored when the block is left (no matter how it's left).

It's very useful for exactly this use case

Reply:
Ahh yep, thanks!  I've just updated this function to use local instead of save/restore